Feature: Double click expand/collapse treeview and numpad keyboard shortcuts

Glen-Nicol-Garmin opened the pull request:
## What does the pull request do?
This pull request adds double click handling to the header of TreeViewItems that will toggle expansion. It also adds common keyboard shortcuts for tree views. 
* numpad * -> Expand all
* numpad / -> collapse all
* numpad + -> expand current item (same as right arrow)
* numpad -  -> collapse current item (same as left arrow)


## What is the updated/expected behavior with this PR?
The behavior is described above and is easily observed using the TreeViewPage in the control catalog. Also in the dev tools window.


## Checklist

- [x ] Added unit tests (if possible)?
- [x ] Added XML documentation to any related classes?
- [- ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
None

## Obsoletions / Deprecations
None

## Fixed issues
None

## Legal
I am contributing this on behalf of Garmin ©. Our legal team needs me to disclose that we are contributing this code under the [MIT license](https://github.com/AvaloniaUI/Avalonia/blob/2ab5f8b8932e0acac078a58c30ff8db908d9a9d8/licence.md) that this project used at the time of contribution.

Glen-Nicol-Garmin committed:
Feature: Double click expand/collapse treeview and numpad keyboard shortcuts

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0026948-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

robloo commented:
> numpad * -> Expand all
> numpad / -> collapse all

Where did this idea originate? Is it a convention used in other controls or frameworks? Personally, this doesn't seem intuitive at all. +/- do make some sense though. 

Would it make sense to just do everything through the enter/space key? Or the Arrow key which are also on the numpad if you disable num lock?

| State | Key Press | New State |
|-|-|-|
| Collapsed Node | Enter -or- Space | Expanded Node |
| Collapsed Node | Ctrl + Enter -or- Ctrl + Space | Expanded Branch (all) |
| Expanded Node | Enter -or- Space | Collapsed Node |
| Expanded Node | Ctrl + Enter -or- Ctrl + Space | Collapsed Branch (all) |
| Collapsed Node | Right Arrow | Expanded Node |
| Collapsed Node | Ctrl + Right Arrow | Expanded Branch (all) |
| Expanded Node | Left Arrow | Collapsed Node |
| Expanded Node | Ctrl + Left Arrow | Collapsed Branch (all) |

 1. Note how a key press -- either Enter or Space -- simply toggles the state. Adding the Ctrl modifier would expand or collapse the entire branch instead of the node.
 2. Arrows (most intuitive) can also be used to expand or collapse the entire branch using the Ctrl modifier.
 3. It is also possible to use the arrow keys on the number pad

robloo commented on src/Avalonia.Controls/TreeView.cs:
```diff
@@ -179,6 +179,27 @@ public void ExpandSubTree(TreeViewItem item)
 
+        /// <summary>
+        /// Collapse the specified <see cref="TreeViewItem"/> all descendent
+        /// <see cref="TreeViewItem"/> s.
```
nit: space probably should be before the last ` s`

Glen-Nicol-Garmin replied:
yeah sorry, I have a linter that auto formats those sometimes.

Glen-Nicol-Garmin replied:
Addressed.

robloo commented on src/Avalonia.Controls/TreeViewItem.cs:
```diff
@@ -198,8 +255,20 @@ protected override void OnKeyDown(KeyEventArgs e)
+                previousInputMethod.DoubleTapped -= HeaderDoubleTapped;
+            }
+
+            var h = e.NameScope.Find<Control>("PART_Header");
```
There doesn't seem to be any need for a new variable `h` here. I would just set directly to `_header` which is easier to read and less verbose.

Glen-Nicol-Garmin replied:
I can do that. I am just so used to that pattern for multithreaded programs when dealing with mutable properties.

robloo replied:
Well, in this case you are getting a reference to a control for event subscriptions and other internal usages. All of that code is sharing the same reference which doesn't change until the template is re-applied. So it is quite a bit different of a case than accessing a potentially mutable property.

If you were getting a property value out of the property system, yes, it's common to do that once at the start of the method. This is usually for performance though. There is only one UIThread in Avalonia right now -- although the new composition thread exists too. Anyway, you don't have to worry most of the time about state change in a single method in a control -- everything is running on the same UIThread.

Glen-Nicol-Garmin replied:
Addressed

robloo commented on tests/Avalonia.UnitTests/MouseTestHelper.cs:
```diff
@@ -104,7 +104,18 @@ public void Move(Interactive target, Interactive source, in Point position, KeyM
+        public void DoubleClick(Interactive target, MouseButton button = MouseButton.Left, Point position = default,
+            KeyModifiers modifiers = default)
+            => DoubleClick(target, target, button, position, modifiers);
+        public void DoubleClick(Interactive target, Interactive source, MouseButton button = MouseButton.Left,
```
nit: Missing blank line above

Glen-Nicol-Garmin replied:
I did this to follow the existing Click methods. Would you prefer I fix that spacing too in this PR?

robloo replied:
Personally, I would always add the blank lines or the vast majority of standard formatting rules throw warnings. Usually, I make small formatting fixes to related areas of code together. Everyone is different though and there are special cases.

Glen-Nicol-Garmin replied:
Addressed, and fixed other lines in this file.

robloo replied:
Great!

Glen-Nicol-Garmin commented:
> Where did this idea originate?

I think the * shortcut has been part of windows tree views for a long time. Windows Explorer is an easy example. Regedit is another easy tree view to get to. Visual studio uses the * as well for its tree views. Admittedly I guess / is not built into windows or visual studio. So I guess that is more for completeness. Ironically, I find myself needing collapse all more than expand all.

The +/- are also windows shortcuts for treeviews.

I think it is would ideal to follow at least the */+/- conventions from windows if we are adding more keyboard shortcuts. But I am open to adding enter and space for toggle current item in addition to the +/- operations. 

I like the idea of using Ctrl or shift to enter recursive mode. But I do have some questions given my preference for keeping */+/-. 

1. Would Ctrl + enter/space toggle recursively? or would it toggle the current item and then recursively set the subtree to match?
2. Would it be too weird to have * and Ctrl + "+" ?
3. Since / is actually not a pre-existing known shortcut would you prefer I make it Ctrl + *? I find that a bit weird if we use your idea to apply the operation recursively. I think I like using / because it completes the set of operations without needing ctrl and allows us to match the windows * behavior.

robloo commented:
> I think the * shortcut has been part of windows tree views for a long time. Windows Explorer is an easy example. Regedit is another easy tree view to get to.

Ah, ok, if there is convention I have less of a concern. I only did a quick check of WinUI where this doesn't seem to be supported. The only potential concern is if this is so windows specific it would be strange on other platforms. I don't see much harm though.

> Would Ctrl + enter/space toggle recursively? or would it toggle the current item and then recursively set the subtree to match?

I was thinking it would toggle the currently focused node -- and then recursively set the same expanded/collapsed state to all the children matching the first parent node. It's a true expand all, collapse all for the branch. So the second one. However, this is just an idea.

> Would it be too weird to have * and Ctrl + "+" ?

Ctrl + "+" is probably a bit strange, yes. Ctrl + Enter/Space is what I was thinking but I realize if we support many keys for expanding/collapsing the Ctrl modifier should be consistent.

> Since / is actually not a pre-existing known shortcut would you prefer I make it Ctrl + *? I find that a bit weird if we use your idea to apply the operation recursively. I think I like using / because it completes the set of operations without needing ctrl and allows us to match the windows * behavior.

I agree if `*` stays it makes more sense to support `\` to collapse all. The operation wouldn't be recursive but would apply the same parent node state to all children as clarified above. 

Since `*` is a convention on Windows and `\` is an obvious continuation of that idea to collapse all I don't have major concerns. I was just unaware Windows allowed this in the old Win32 controls. It's probably best to wait and see what others say here. If everyone is OK with it and it works for your use case there probably isn't' a need to add my toggle ideas at this time.

grokys commented:
> The only potential concern is if this is so windows specific it would be strange on other platforms.

I haven't had time to look at this PR in depth yet (though looks good from a quick glance!) but we have this problem on various controls to be honest so I'm OK with adding windows-specific interactions here. We may want, in future, to refactor interaction handling for some controls into a separate handler as we do with `IMenuInteractionHander` in order to be able to switch if by platform, but that would be something for the future.

Glen-Nicol-Garmin commented:
So this is now the current keys

| Key Combo | Action |
|--------------| --------|
| Arrow left | collapse|
| Arrow right | expand |
| Space | toggle |
| Enter | toggle |
| Ctrl+ Arrow left | collapse recursively|
| Ctrl + Arrow right | expand recursively|
| Ctrl + Space | recursively toggle based on focused item |
| Ctrl + Enter | recursively toggle based on focused item |
| numpad - | collapse |
| numpad + | expand |
| numpad /  | recursively collapse |
| numpad * | recursively expand |

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0027086-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

robloo commented on src/Avalonia.Controls/TreeViewItem.cs:
```diff
@@ -166,30 +168,95 @@ protected override void OnKeyDown(KeyEventArgs e)
+                Func<TreeViewItem, bool>? handler =
+                    e.Key switch
+                    {
+                        Key.Left => ApplyRecursivelyIfCtrl(FocusAwareCollapseItem, e.KeyModifiers),
+                        Key.Right => ApplyRecursivelyIfCtrl(ExpandItem, e.KeyModifiers),
```
I don't see where only Left/Right is handled to expand/collapse the focused node. This was the old functionality. The new logic only applies if the Ctrl key is held and I don't see anywhere else where Left/Right keys are used. Am I missing something?

robloo replied:
Nevermind, I see how this works now... I don't know why I can't mark as resolved though since I created this comment. Feel free to close this and others to cleanup the comments.

Glen-Nicol-Garmin replied:
Should I rename the function to make that clearer that it does the actual and does it recursively if ctrl is held?

robloo replied:
The renamed function definitely makes it more clear, thanks! Local function naming really is up to the dev. It doesn't affect anything else so whatever you think best.

robloo commented on src/Avalonia.Controls/TreeViewItem.cs:
```diff
@@ -166,30 +168,95 @@ protected override void OnKeyDown(KeyEventArgs e)
+                        _ => null,
+                    };
+
+                if(handler is not null)
```
nit: `if (` should have a space

Glen-Nicol-Garmin replied:
doh, revenge of the linter for being turned off.

robloo commented on src/Avalonia.Controls/TreeViewItem.cs:
nit: `if (` should have a space

robloo commented on src/Avalonia.Controls/TreeViewItem.cs:
```diff
@@ -166,30 +168,95 @@ protected override void OnKeyDown(KeyEventArgs e)
+                    e.Handled = handler(this);
+                }
+
+
```
nit: Extra blank line here.

Feel free to ignore some of these nits. Avalonia tries to maintain high code formatting quality but it can get monotonous and it's better to just accept new functionality rather than make it perfect.

robloo commented:
Thanks for adding more key shortcuts!

Glen-Nicol-Garmin commented:
I addressed the nits and renamed that function to make it more obvious that it does the action and recursively if Ctrl.

One thing I did notice that seems independent of my changes. The tree view seems to leak memory. If you run treeview page on the control catalog and hit the star key 4 times for one of the root elements that will quickly expand that item to many thousands of items (10 ^ 4 I think?). Which raises the apps memory usage to 1.1 GB. Understandable. But! then when you hit the remove button, the memory actually goes up further! and then doesn't seem to ever come down in a meaningful way. I ruled out my new double click handler by commenting that out to make sure that wasn't holding on to them. The memory snapshots I took seemed to imply it may have been the item container generator that was holding the references but that is not for sure.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0027195-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

Glen-Nicol-Garmin commented:
Requested changes have been addressed and resolved. Thanks for the review and time to think about other keys! Seems the PR just needs approval?

grokys commented:
> One thing I did notice that seems independent of my changes. The tree view seems to leak memory. If you run treeview page on the control catalog and hit the star key 4 times for one of the root elements that will quickly expand that item to many thousands of items (10 ^ 4 I think?). Which raises the apps memory usage to 1.1 GB. Understandable. But! then when you hit the remove button, the memory actually goes up further! and then doesn't seem to ever come down in a meaningful way.

Could you open an issue for this? I'm currently refactoring `ItemsControl`s including `TreeView` and will address this.

grokys commented:
Thanks for this! Approved.

grokys committed:
Merge branch 'master' into treeview_double_click

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0027321-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

Glen-Nicol-Garmin commented:
Sure thing, #9662 

Thanks for a pleasant and helpful PR experience!
